[RISCV] Mark VFIRST and VCPOP as SignExtendingOpW

mshockwave opened the pull request:
Since their values are small enough ([-1, 65535] & [0, 65535], respectively) to fit into signed 32 bits, any sext (or downcasting + sext) will be redundnat. Hence marking them as SignExtendingOpW.

topperc commented on llvm/test/CodeGen/RISCV/opt-w-instrs.mir:
```diff
@@ -28,3 +28,55 @@ body:             |
+     %0:vr = COPY $v8
+     %1:gprnox0 = COPY $x10
+     %2:gpr = PseudoVFIRST_M_B1 %0:vr, %1:gprnox0, 0
+     %3:gpr = ADD %1, %2
```
What is the ADD instruction for? I see it was in the fcvt test too but I can't recall why

mshockwave replied:
IIRC we need an additional use of `%2` otherwise `ADDI` will be removed before RISCVOptWInstrs even kicks in. Let me double check.

mshockwave replied:
If the user of sext only uses the lower 32 bits, RISCVOptWInstrs will remove the sext regardless of the SignExtendingOpW flag (see the `hasAllWUsers` function). Previously, I tried to use `ADD` to create that artificial use, despite using an incorrect syntax (`ADD` was using `%2` not `%4`) it somehow stoped RISCVOptWInstrs from removing sext in absent of SignExtendingOpW.
But even `ADD` is not the best choice as RISCVOptWInstrs transitively looks into the user of `ADD` to determine if only the lower 32 bits are used so we're only pushing off the problem. A better choice would be using `SRLI` with its shift amount greater than 32, which creates an artificial use beyond 32 bits. I'll update the patch accordingly

topperc replied:
I'm confused. The ADD and ADDIW were separate users of the VFIRST. The ADDIW had no users other than a COPY . A COPY should be assumed to use all bits. So how was it allowed to be removed?

mshockwave committed:
Use SRLI to create additional use of sext values

mshockwave committed:
[RISCV] Mark VFIRST and VCPOP as SignExtendingOpW

Since their values are small enough ([-1, 65535] & [0, 65535],
respectively) to fit into signed 32 bits, any sext (or downcasting +
sext) will be redundnat. Hence marking it as SignExtendingOpW.

mshockwave committed:
Use COPY to create artificial uses

mshockwave commented:
I've rebased to leverage COPY to create artificial (64-bit) uses enabled by 4dd5d967975fa8d52b8c60596d892d9dd5615809

topperc reviewed (approved):
LGTM
